internal/permissions/groups/middleware.go

```diff
@@ -0,0 +1,90 @@
+package userpermissions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+
+	"github.com/go-chi/chi"
+	"github.com/sirupsen/logrus"
+	"github.com/source-academy/stories-backend/internal/auth"
+	"github.com/source-academy/stories-backend/internal/database"
+	groupenums "github.com/source-academy/stories-backend/internal/enums/groups"
+	apierrors "github.com/source-academy/stories-backend/internal/errors"
+	"github.com/source-academy/stories-backend/model"
+)
+
+type contextKey string
+
+const (
+	GroupKey contextKey = "group_context"
+	RoleKey  contextKey = "role_context"
```

Comment:
Would be better if these keys were private (i.e. `groupKey` and `roleKey`) to prevent misuse.

Will approve for now but please refactor in the follow-up PR.